[DO NOT SQUASH] KPACK logic.

whchung opened the pull request:
KPACK logic.

Divide GemmK dimension into GemmK and GemmKPack dimension to allow wider LDS -> VGPR data transfer.

Features:
- supports fp16 / fp32.
- supports NCHW / NHWC.
- could be tuned via `--perf_config`.
- forward and backward convolutions for Resnet50 configs are all validated.

Limitations:
- XDLOPS path only. The logic does support non-XDLOPS path but is not enabled yet.
- forward convolution only. does NOT support backward convolutions yet.
- does NOT support padding kernels along K/M/N directions.

whchung committed:
Introduce KPack into conv2d lowering logic.

whchung committed:
Aggregate arg0TargetLayoutNameX variables into a vector.

whchung committed:
Aggregate arg1TargetLayoutNameX variables into a vector.

whchung committed:
Aggregate arg2TargetLayoutNameX variables into a vector.

whchung committed:
Consider gemmKExtra while computing KPACK.

whchung committed:
Explicitly use Value type for the result of miopen.transform ops.

whchung committed:
Initial commit to do KPack handling for filter tensor in forward convolution.

whchung committed:
Initial commit to do KPack handling for input tensor in forward convolution.

whchung committed:
Supply KPACK as an attribute to gridwise_gemm and gridwise_gemm_v2.

whchung committed:
Fix bugs to get past -miopen-affine-transform pass.

whchung committed:
Obtain KPack attribute in gridwise_gemm and gridwise_gemm_v2 lowering logic.

whchung committed:
XXX WIP lowering logic of miopen-lowering-step2 for non-XDLOPS path.

whchung committed:
XXX WIP lowering logic of miopen-lowering-step3 for non-XDLOPS path.

whchung committed:
Fix clang-format for the previous couple of commits.

whchung committed:
XXX WIP lowering logic of miopen-lowering-step3 for non-XDLOPS path.

DONE:

- miopen.threadwise_gemm

TBD:

- miopen.threadwise_load
- miopen.threadwise_store
- miopen.threadwise_copy

whchung committed:
XXX WIP supply KPack information to blockwise_copy.

whchung committed:
XXX WIP lowering logic of miopen-lowering-step4 for non-XDLOPS path.

CAVEATS:

- miopen.threadwise_load / miopen.threadwise_store have to use legacyLoad / legacyStore path.

whchung committed:
Fix typo.

whchung committed:
Fix clang-format.

whchung committed:
Suppress debug messages for non-XDLOPS path.

whchung committed:
XXX WIP lowering logic of miopen-lowering-step2 for non-XDLOPS path.

Slice length for GemmK dimension for matrix A / B needs to be reviewed for both non-XDLOPS and XDLOPS path when vector load dimension is on GemmM or GemmN.

whchung committed:
XXX WIP commit to get XDLOPS path working.

whchung committed:
Get rid of unused codes.

whchung committed:
Pass KPack to xdlops_gemm_v2 and change VGPR types to hold for Matrix A / B to vectors.

whchung committed:
XXX WIP. Lowering logic for xdlops_gemm_v2 considering KPACK.

whchung committed:
Suppress debug messages.

whchung committed:
Fix LDS allocation size.

whchung committed:
Fix VGPR allocation size for LDS->VGPR transfer.

whchung committed:
Fix LDS->VGPR load loop iteration.

whchung committed:
Fix data copied per thread.

whchung committed:
Temp commit for some debug logic.

whchung committed:
Fix a bug in KReduction path.

whchung committed:
Remove unused codes.

whchung committed:
Suppress debug messages.

whchung committed:
Revert "Remove the undesirable LDS barrier."

This reverts commit 5aa2c6746d8fb41767d83bd13917dbe90919ca93.

whchung committed:
Add one set of default tuning parameters.

whchung committed:
Fix coordinate computation to LDS.

whchung committed:
Fix coordinate computation on global loads.

whchung committed:
Tweak slice lengths for Matrix B.

whchung committed:
Change UnMerge attribute from "dimension_lengths" to "parameters".

Cope with requirements imposed in computeIndexDiffMap logic.

whchung committed:
Re-enable index diff map for KPACK logic.

whchung committed:
XXX. First end-to-end working example with -rand=1.

whchung committed:
Suppress debug messages.

whchung committed:
Comment out unused variable.

whchung committed:
Move the location of LDS barrier to reduce number of s_barriers issued.

whchung committed:
Add comment on a weird tuning parameter.

whchung committed:
Revert "Move the location of LDS barrier to reduce number of s_barriers issued."

This reverts commit 67e2cc8eb2aa3569be30b5371df795cf6c742d6a.

whchung committed:
Move LDS barrier positions.

whchung committed:
Remove the comment of the tuning parameter.

whchung committed:
XXX. Initial logic to get KPACK=8 and KPACK=4 working.

whchung committed:
Add one additional tuning parameters.

whchung committed:
Refactor logic.

whchung committed:
Fix the logic to support original f32 lowering logic.

whchung committed:
Let f32 type use KPACK as well.

whchung committed:
Remove obsolete comments.

whchung committed:
Hard-code initial KPack value as 4 for f16.

whchung committed:
Improve MFMA argA/B loading logic to support all 4 possible cases.

whchung committed:
Add comments for the new MFMA code emission logic in !KReduction path.

whchung committed:
Fix clang-format.

whchung committed:
Set kpack tuning parameter in -miopen-affix-params.

whchung committed:
Add guarding logic to allow step2 fail gracefully.

whchung committed:
Fix E2E tests for NHWC layout.

whchung committed:
Fix lowering failures for f16 backward weight convolution.

whchung committed:
Fix two unit tests.

whchung committed:
Fix 3 unit tests. Disable KPACK in case we need to use padding kernel.

  MLIR :: mlir-miopen-driver/auto_e2e/padding_kernel_all_fwd.mlir
  MLIR :: mlir-miopen-driver/auto_e2e/padding_kernel_all_wrw.mlir
  MLIR :: mlir-miopen-driver/auto_e2e/padding_kernel_gemmK.mlir

whchung committed:
XXX FIXME: temporary band-aid to disable failing configs.

whchung committed:
Merge branch 'miopen-dialect' into miopen-dialect-kpack-logic

whchung committed:
Remove some compile-time warnings.

whchung commented:
We can NOT trust the green check mark just yet as I disabled quite a few tests on commit 629e309. The next steps for me is to address all those disabled tests in that commit.

whchung commented:
Attaching skip-ci label for now. And convert it to draft.

Will undo all these after I fix some more e2e unit tests.

whchung committed:
TypeRange -> Type.

whchung committed:
Re-arrange Resnet50 e2e tests.

whchung committed:
Fix various f16 unit tests by introducing KForOuterLoop when KPack is 1.

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:
```diff
@@ -103,9 +103,9 @@ def MIOpen_TransformOp :
-                   MemRefRankOf<[F32, F16, I16], [3]>:$input,
-                   MemRefRankOf<[F32, F16, I16], [3]>:$output)> {
+    Arguments<(ins MemRefRankOf<[F32, F16, I16], [3, 4]>:$filter,
+                   MemRefRankOf<[F32, F16, I16], [3, 4]>:$input,
```
I take it we'll be replacing this with the variadic arguments?

krzysz00 commented on mlir/test/mlir-miopen-driver/auto_e2e/Resnet50/Resnet50_nchw_f16_bwd.mlir:
```diff
@@ -1,3 +1,6 @@
+// XFAIL: *
+// FIXME, ALL CONFIGS IN THIS TEST FILE NEED TO BE REVIEWED.
```
Reviewed in what sense?

krzysz00 commented on mlir/test/mlir-miopen-driver/auto_e2e/padding_kernel_all_fwd.mlir:
```diff
@@ -3,7 +3,7 @@
 // CHECK_RESNET50_CONFIG1: [1]
 
-// RUN: mlir-miopen-driver --operation=conv2d -t f16 -p=false  -fil_layout=gkcyx -in_layout=ngchw -out_layout=ngkhw -batchsize=20 -groupsize=1 -in_channels=3 -out_channels=6 -in_h=32 -in_w=32 -fil_h=7 -fil_w=7 --dilation_h=1 --dilation_w=1 --padding_h=3 --padding_w=3 --conv_stride_h=2 --conv_stride_w=2 %pv %random_data %xdlops -c | mlir-rocm-runner --shared-libs=%rocm_wrapper_library_dir/librocm- ...
+// NOT: mlir-miopen-driver --operation=conv2d -t f16 -p=false  -fil_layout=gkcyx -in_layout=ngchw -out_layout=ngkhw -batchsize=20 -groupsize=1 -in_channels=3 -out_channels=6 -in_h=32 -in_w=32 -fil_h=7 -fil_w=7 --dilation_h=1 --dilation_w=1 --padding_h=3 --padding_w=3 --conv_stride_h=2 --conv_stride_w=2 %pv %random_data %xdlops -c | mlir-rocm-runner --shared-libs=%rocm_wrapper_library_dir/librocm- ...
```
Why're we flipping tests off?

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -770,15 +776,30 @@ inline void emitNaiveTensorCopyLogic(
-      srcUpperIndices.push_back(b.create<IndexCastOp>(
-          loc, b.create<AddIOp>(loc, ivi_i32, sourceCoord[2]),
-          b.getIndexType()));
+      if (sourceCoord.size() == 3) {
```
This is all supersceded by #425, sorry about the merge conflicts

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -1767,19 +1810,19 @@ struct Conv2DRewritePattern : public OpRewritePattern<T> {
                          b.getArrayAttr({b.getI32IntegerAttr(1)})),
           b.getNamedAttr("upper_layer_names", b.getArrayAttr({b.getStringAttr(
-                                                  arg0TargetLayoutName1)}))};
+                                                  arg0TargetLayoutName[1])}))};
```
Nice little bit of cleanup here!

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -1829,9 +1870,9 @@ struct Conv2DRewritePattern : public OpRewritePattern<T> {
-        b.getArrayAttr({b.getStringAttr(arg0TargetLayoutName0),
-                        b.getStringAttr(arg0TargetLayoutName1),
-                        b.getStringAttr(arg0TargetLayoutName2)})));
+        b.getArrayAttr({b.getStringAttr(arg0TargetLayoutName[0]),
```
Could this instead be `b.getStrArrayAttr()`?

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -2906,16 +3045,113 @@ struct Conv2DRewritePattern : public OpRewritePattern<T> {
       // input padding end
     }
 
+    // KPack for input tensor.
```
Note to our future selves: there might be refactoring possibilities here.

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -5079,18 +5331,38 @@ void affixBlockwiseCopyAttributes(
+                                                b.getI32IntegerAttr(1),
+                                                b.getI32IntegerAttr(2),
+                                                b.getI32IntegerAttr(3),
+                                            }));
```
`b.getI32ArrayAttr({0, 1, 2 , 3})`

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -5390,85 +5672,152 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<miopen::GridwiseGemm
-          GemmABlockCopyNumberDataPerThread /
-          GemmABlockCopyThreadSliceLengths_GemmM;
+      // TBD: FIXME. Review logic here.
+      if (KPack > 1) {
```
Something doesn't feel right about the logic here. Can you talk me through what these variables do, how they're computed, and how they're affected by kpack?

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -5484,28 +5833,110 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<miopen::GridwiseGemm
+    GemmABlockCopyDestCoord_Y_i32 =
         b.create<AddIOp>(loc, zeroConstantI32Op, GemmAThreadDataIdBegin_Y_i32);
-    auto GemmABlockCopyDestCoord_X_i32 =
+    GemmABlockCopyDestCoord_X_i32 =
```
Can we get rid of these +0 ops? They'll make the code clearer and might enable more optimizations (see #440).

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -5613,9 +6143,15 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<miopen::GridwiseGemm
-    SmallVector<uint64_t, 3> blockwiseCopyABounds = {
-        1, GemmABlockCopyThreadSliceLengths_GemmK,
-        GemmABlockCopyThreadSliceLengths_GemmM};
+    SmallVector<uint64_t, 3> blockwiseCopyABounds;
```
Let's update this to 4 to prevent heap allocations

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -5634,17 +6172,23 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<miopen::GridwiseGemm
-        1, GemmBBlockCopyThreadSliceLengths_GemmK,
-        GemmBBlockCopyThreadSliceLengths_GemmN};
-    TypeRange blockwiseLoadBType;
+    SmallVector<uint64_t, 3> blockwiseCopyBBounds;
```
Same size update here

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -5676,58 +6222,126 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<miopen::GridwiseGemm
+      blockwiseCopyVectorCoordType = VectorType::get({3}, b.getIntegerType(32));
+    }
 
     // Matrix A: {0, 0, m_block_data_on_global}, {0, 0, 0}
```
Let's update this comment for kpack

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -5849,19 +6463,28 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<miopen::GridwiseGemm
+
+    Value blockwiseCopyASrcVectorUpdated;
+    if (KPack > 1) {
+      blockwiseCopyASrcVectorUpdated = lb.create<miopen::MovePosV2Op>(
```
This entire section of the change'll get clobbered by #425.

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -6447,79 +7055,152 @@ struct GridwiseGemmV2RewritePattern
     // Compute ThreadSliceLengths for Matrix A.
     uint64_t GemmABlockCopyNumberDataPerThread =
-        MPerBlock * KPerBlock / BlockSize;
+        MPerBlock * KPerBlock * KPack / BlockSize;
```
Let's put some parentheses around the multiplication to make sure we don't get messed up by `KPack / BlockSize` ending up as part of the computation (for all I know, it might)

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -6447,79 +7055,152 @@ struct GridwiseGemmV2RewritePattern
-          GemmBBlockCopyNumberDataPerThread /
-          GemmBBlockCopyThreadSliceLengths_GemmN;
+      // TBD: FIXME. Review logic here.
+      if (KPack > 1) {
```
As a general thought, would anything break (especially given #440) if, instead of all these `if KPack > 1` checks, we just always had a KPack dimension that was sometimes fixed to 1?

krzysz00 replied:
(and yeah, I know that applies to #441's approach to the output tensor as well)

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -6565,28 +7327,108 @@ struct GridwiseGemmV2RewritePattern
+    if (KPack > 1) {
+      GemmBBlockCopyThreadClusterId_Z = b.create<SignedDivIOp>(
+          loc,
+          b.create<SignedDivIOp>(loc, tid,
```
Signed division? Neither of these can be negative, let's go unsigned instead

krzysz00 replied:
Heck, that goes for all the rest of these computations.

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -6565,28 +7327,108 @@ struct GridwiseGemmV2RewritePattern
+    Value GemmBBlockCopySourceCoord_Z_i32;
+    if (KPack > 1) {
+      GemmBBlockCopySourceCoord_Z_i32 = b.create<AddIOp>(
+          loc, zeroConstantI32Op, GemmBThreadDataIdBegin_Z_i32);
```
Same as above with adding 0

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -8232,45 +9260,97 @@ struct ThreadwiseGemmRewritePattern
+      Value add;
+      if (dataType.isa<IntegerType>())
+        add = b.create<AddIOp>(loc, dataType, mul, gemmCMN);
+      else
```
Braces on all the if statements, please? It makes adding debugging info easier, and I'm tempted to ask we add a lint forbidding this sort of thing going forward.

whchung replied:
done

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -8416,6 +9496,7 @@ struct ThreadwiseCopyRewritePattern
       ArrayAttr layeredSourceTransformMetadata;
       ArrayAttr layeredDestTransformMetadata;
 
+      legacyLoad = true;
```
Please explain this

whchung replied:
This logic has been eliminated in the latest logic.

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -9883,6 +10974,11 @@ struct XdlopsGemmV2RewritePattern
                   loc, ilnkb.create<MulIOp>(loc, ilnkiv, NConstantOp), laneId),
               ilnkb.create<MulIOp>(loc, NPerXdlopsConstantOp, olniv)));
 
+      if (KPack > 1)
```
Why's this multiplication called for?

whchung replied:
With KPack, we are moving along LDS with stride in the multiples of KPack.

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -9911,48 +11015,133 @@ struct XdlopsGemmV2RewritePattern
+          assert(bufferBElementType.isa<VectorType>());
+
+          argA = innerLoopb.create<SplatOp>(loc, zeroOp, argType);
+          argB = innerLoopb.create<SplatOp>(loc, zeroOp, argType);
```
We have zero constant creation functions in #425 (and there's a small fix to them in #441 I might want to backport related to bfloats) - FYI for your merge/rebase

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -10010,8 +11199,12 @@ struct XdlopsGemmV2RewritePattern
-                                        ValueRange{sourceOffsetA});
+      Value valueA;
+      if (KPack > 1) {
+        valueA = emitLoadLogic(lklb, loc, op.matrixA().getType().template cast<MemRefType>(), argType, false, {}, op.matrixA(), ValueRange{sourceOffsetA});
```
General comment: is it possible to statically compute the operations emitLoadLogic will lead to and apply those transformations to sourceOffsetA?

whchung replied:
the computation depends on `waveOffsetA`, which is derived from the current workitem ID, so it's not statically determined.

krzysz00 reviewed (commented):
Ok, so, this PR is pretty noisy due to how copy-pasted our lowering code is, so it's hard to give high-level feedback, but I've left some style notes, a few questions, and notes on things that'll get clearer (or at least messed up by) #425  (which really had rough timing).

The logic you marked with a TBD confused me as well - it felt like the divisions were in the wrong order, so more words about it would be nice.

whchung committed:
Add back fixed padding kernel e2e tests.

whchung committed:
Storing current status of e2e unit tests.

- PASS: those cases which would pass.
- NOT: those cases which would fail.

Each of the test file still contains at least 1 RUN config.

whchung committed:
Remove XFAIL from 2 e2e tests as they now all pass.

whchung committed:
Fix two lowering unit tests.

whchung committed:
Remove XFAIL from 2 e2e tests as they now all pass.

whchung committed:
Revise initial tuning parameters and fix all failing e2e f32 tests.

whchung committed:
Reinstate passing e2e f16 forward tests.

Cases still failing within the current tuning parameter set are marked
as FIXME.

whchung committed:
Reintroduce index diff maps load logic.

whchung committed:
XXX. Mark 1 e2e test module as XFAIL for now.

whchung committed:
Fix an uninitialized variable.

whchung committed:
Fix quite a few f16 backward convolution e2e unit tests.

Still some of them are not yet fixed. Working on it.

whchung committed:
Fix all remaining failing tests in conv2d_host_validation_f16_bwd.mlir

whchung committed:
Fix clang-format.

whchung committed:
Use different set of initial tuning parameters between forward and backward convolutions.

whchung committed:
Merge remote-tracking branch 'origin/miopen-dialect' into miopen-dialect-kpack-logic

whchung commented:
jenkins: retest this please.

whchung committed:
Fix clang-format.

whchung committed:
Fix two unit tests which are missing CHECK macros.

whchung committed:
Fix bf16 lowering pipeline.

whchung committed:
Fix offset computation logic for both KPACK and non-KPACK situtation.

whchung commented:
jenkins: retest this please.

whchung committed:
Suppress debug messages.

whchung committed:
Fix one e2e test lit check macro.

whchung committed:
Fix clang-format.

whchung committed:
Address review comments adding braces.

whchung commented:
jenkins: retest this please.

whchung committed:
Disable tuning when KPACK is in use.

This is a temporary measure.

whchung committed:
Merge branch 'miopen-dialect' into miopen-dialect-kpack-logic

whchung commented:
Addded @yiqian1 as a reviewer. It has been identified with @jerryyin and I that existing KPACK logic would introduce butterfly effect that the FASTEST kernel coming from the tuning parameter search space currently used by MIOpen MLIR solver would result in INCORRECT kernel. On the other hand, the kernel is working fine if using hand-picked heuristic tuning parameter currently employed in the branch.

Before the root cause is yet known, disabling tuning for now.

yiqian1 commented:
> Addded @yiqian1 as a reviewer. It has been identified with @jerryyin and I that existing KPACK logic would introduce butterfly effect that the FASTEST kernel coming from the tuning parameter search space currently used by MIOpen MLIR solver would result in INCORRECT kernel. On the other hand, the kernel is working fine if using hand-picked heuristic tuning parameter currently employed in the branch.
> 
> Before the root cause is yet known, disabling tuning for now.

I'll disable the weekly CI once this PR is merged. It doesn't make sense to run weeklies with tuning disabled.

krzysz00 commented:
@whchung I'd prefer that we debug why KPACK doesn't work with all tuning parameters or to add a flag (like my `enableOutSwizzles`) that disables KPACK when the tuning parameters cannot accomodate KPACK

krzysz00 commented on external/llvm-project/mlir/.clang-format:
```diff
@@ -1,2 +1,2 @@
 BasedOnStyle: LLVM
-AlwaysBreakTemplateDeclarations: Yes
\ No newline at end of file
+AlwaysBreakTemplateDeclarations: true
```
Why'd we make this change?

whchung replied:
Some places in LLVM upstream use `Yes`, and some use `true`. I seem to bump into some issues locally awhile so I introduced this line. Will remove it.

krzysz00 replied:
We can probably keep it - interestingly, upstream doesn't seem to set this value at all?
I just was trying to understand what the motivation here was.

whchung commented:
@krzysz00 The bug is in several different layers beyond the scope of MIOpen dialect so it's a bit nasty. I'm still in the process to figure out where in the parameter search space triggered the issues.

- vector dialect lowering : complete legit `vector.transfer_read` op ended up crashing at lowering to LLVM dialect. This is what causing weekly CI to break.
- backend : complete legit tuning parameters ending up with incorrect kernel. I forced setting KPACK to be all 1 from MIOpen and started tuning. And even with that we get incorrect kernels, and unfortunately that incorrect one happens to be the fastest one.

krzysz00 commented:
@whchung Yowch. Good luck and let me know if you need a sounding board!

whchung commented:
Now if I could get rid of the vector -> llvm lowering failure we could have a passing weekly CI...

asroy commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -5265,7 +5517,7 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<miopen::GridwiseGemm
         math_util::integer_divide_ceil<int64_t>(MPerBlock, max_lds_align);
     a_block_space = math_util::integer_least_multiple(KPerBlock * AlignedMPerBlock,
-                                                 max_lds_align);
+                                                 max_lds_align) * KPack;
```
This is the C++ logic for A/B tensor in LDS with [KPerBlock, MPerBlock, KPack] layout

https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/a4f24233e51854c4b5cb7d75637fa0f235f78f8e/composable_kernel/include/tensor_operation/gridwise_gemm_xdlops_v2r3.hpp#L409-L439

asroy replied:
a_block_space = math_util::integer_least_multiple(KPerBlock * MPerBlock, max_lds_align) * KPack

same for b matrix

whchung replied:
@asroy . PR #490 .

asroy commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -5265,7 +5517,7 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<miopen::GridwiseGemm
         math_util::integer_divide_ceil<int64_t>(MPerBlock, max_lds_align);
     a_block_space = math_util::integer_least_multiple(KPerBlock * AlignedMPerBlock,
-                                                 max_lds_align);
+                                                 max_lds_align) * KPack;
```
This is the C++ logic for A/B tensor in LDS with [KPerBlock, MPerBlock, KPack] layout

https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/a4f24233e51854c4b5cb7d75637fa0f235f78f8e/composable_kernel/include/tensor_operation/gridwise_gemm_xdlops_v2r3.hpp#L409-L439

whchung committed:
Remove more restrictions.

whchung committed:
Refactor codes.

whchung committed:
XXX Experimental commit.

whchung committed:
Merge remote-tracking branch 'origin/miopen-dialect' into miopen-dialect-kpack-logic

whchung commented:
> @krzysz00 The bug is in several different layers beyond the scope of MIOpen dialect so it's a bit nasty. I'm still in the process to figure out where in the parameter search space triggered the issues.
> 
> * vector dialect lowering : complete legit `vector.transfer_read` op ended up crashing at lowering to LLVM dialect. This is what causing weekly CI to break.

Turns out we may be able to get rid of `vector.transfer_read` completely. Testing this assumption with CI.

> * backend : complete legit tuning parameters ending up with incorrect kernel. I forced setting KPACK to be all 1 from MIOpen and started tuning. And even with that we get incorrect kernels, and unfortunately that incorrect one happens to be the fastest one.

Will address this after we can confirm it's possible to get rid of all all `vector.transfer_read` ops. Also this PR has been merged with the latest `miopen-dialect` which has @pcf000 's upstream merge.

whchung committed:
Revert "XXX Experimental commit."

This reverts commit 006b1640c0a6eb3ea3c3ae4eba4d9708913460c1.

whchung committed:
Revert "Disable tuning when KPACK is in use."

This reverts commit 949752b9483052f4dfc1a0343f68e9456a6d2546.

whchung committed:
Throttle parallel level in the tuning process.

krzysz00 committed:
Use index almost everywhere (#459)

Now that vector supports indexing by index, there's no good reason
to have i32 values in our lowering pipeline. The one exception, for now,
is that the GPU OOB hack still requires i32 indices to work.

whchung committed:
Get rid of ldconfig invocation in tuning to be in sync with mainline.

whchung committed:
Merge branch 'miopen-dialect' into miopen-dialect-kpack-logic

jerryyin commented on mlir/utils/jenkins/miopen-tests/miopen_validate.sh:
```diff
@@ -136,6 +136,7 @@ function clean_miopen_caches() {
 function setup_environment() {
     export MIOPEN_FIND_MODE=1
     export MIOPEN_DRIVER_USE_GPU_REFERENCE=1
+    export MIOPEN_COMPILE_PARALLEL_LEVEL=1
```
This needs to be reverted for final merge: MIOPEN_COMPILE_PARALLEL_LEVEL=1 is a debugging flag (that alters tuning behavior) and should not be used in production.

whchung replied:
This was added in 91e318b as a measure to get rid of errors encountered in the tuning process. Will keep an eye on it.

whchung replied:
Per discussion with @jerryyin , another temporarily measure is to modify the semantic of `miirLowerTuningParams`, which is used by `IsApplicable` in MIOpen MLIR solvers to determine if a tuning parameter is valid, so that `miirLowerTuningParams` would behave exactly as `miirLowerBin` for the whole lowering pipeline.

The rationale why `MIOPEN_COMPILE_PARALLEL_LEVEL=1` was introduced was because it was identified that:

1) without the environment variable, the tuning process would crash upon receiving a failure in getting a kernel binary blob from `miirLowerbin`. 
2) `miirLowerTuningParams` was invoked prior to `miirLowerBin` to help determine whether a tuning parameter is valid. It does NOT run through the whole lowering pipeline.
3) Should anything bad happens in the later part of the lowering pipeline, `miirLowerTuningParams` couldn't catch that.
4) MIOpen, as a client of `miirLowerTuningParams`, upon receiving the a success return value from the function, would assume `miirLowerBin` would definitely yield a working kernel, which is NOT always the case. Something bad might happen in the later part of the lowering pipeline. It could be in MIOpen dialect, could be in LLVM core, could be in LLVM AMDGPU backend.
5) The exception handling behavior inside MIOpen when `MIOPEN_COMPILE_PARALLEL_LEVEL=1` is NOT set is UNDEFINED. And currently it would just crash the tuning process.

whchung replied:
Commit 41b0a81 incorporate this temporary measure.

whchung replied:
I have just reverted the commit. Empirical testing shows the crash would still occur. And indeed it's inside MIOpen, albeit in a different place.

whchung replied:
I'll keep `MIOPEN_COMPILE_PARALLEL_LEVEL=1` in the test script for now.

After getting a deeper understanding of the other issue (tuning for fp16 backward data / weight) then revisit this topic.

jerryyin replied:
Thanks for confirming. As you want to proceed with merging and revisit those topics anyways I think you can just disable fp16 tuning in general as a temporary measure (Either in bash script or in Jenkinsfile). Otherwise we will be under trial-and-error mode for the code path untested (and probably shouldn't be there). Also, if you decide to disable fp16 tuning, please do file ticket to document all issues in independent tickets. This way we have a way to track those issues. 

The concern I had with all the temporary measures that got to merge with this PR is that it may another implicit deadline: the next ROCm release. On the other hand, if you keep the kpack in a separate branch then it will not impact anything in the mainline. Compared to the two, I actually think keeping it in the separate branch will give us more flexibility as it is unknown how much effort is needed to root cause all those issues.

whchung replied:
In the latest implementation, this hack is no longer necessary.

whchung committed:
Merge branch 'miopen-dialect' into miopen-dialect-kpack-logic

whchung commented:
jenkins: retest this please.

whchung committed:
XXX FIXME Disable tuning for backward data and backward weight for fp16 type.

Due to unknown reason, the fastest kernel yield at the tuning process is
NOT correct.
By disabling tuning, the default heuristic tuning parameter in MLIR
MIOpen dialect would be used, which has been verified to produce correct
results.

whchung commented:
Per discussion with @jerryyin , identified a temporarily measure to skip tuning for backward data and backward weight for fp16 data type. Prior to the latest LLVM upgrade, they could be tuned, but now the fastest kernel after tuning would yield incorrect results.

Currently:

- fp32 : fwd / bwd / wrw could all be tuned and yield correct results.
- fp16 : fwd could be tuned and yield correct results.
- fp16 bwd / wrw tuning has been disabled, and rely on the built-in handpicked tuning parameter set in MLIR MIOpen dialect.

krzysz00 commented:
@whchung Can we get a performance comparison on the kernels that have had their tuning disabled to show that kpack being enabled leads to better performance than not having kpack and allowing for tuning?

I'm not comfortable merging this code without, at the very least, data showing that disabling tuning for broad classes of kernels is worth it.

whchung commented:
> @whchung Can we get a performance comparison on the kernels that have had their tuning disabled to show that kpack being enabled leads to better performance than not having kpack and allowing for tuning?
> 
> I'm not comfortable merging this code without, at the very least, data showing that disabling tuning for broad classes of kernels is worth it.

It is yet known which part caused the fastest kernel in fp16 backward (data / weight) convolution become incorrect : tuning search space / dialect implementation / LLVM core / LLVM backend.

The latest commit is a step to try obtain a full execution of nightly + weekly CI so we can get performance comparisons.

krzysz00 commented:
@whchung You don't need a full nightly/weekly run for performance comparisons - you can manually call the benchmarking scripts on 112 to get results for just the kernels of interests and generate the plots.

Also, re the paralellism level, wouldn't it make more sense to,  in `miirLowerBin` and its friends, have a thing that'll catch exceptions (and _maybe_ signals, though we could just compile with assertions off for release and be under the assumption that we won't segfault) and return failure if an exception occurs? I'm pretty sure that's a typical pattern when wrapping C++ for C?

whchung commented:
> @whchung You don't need a full nightly/weekly run for performance comparisons - you can manually call the benchmarking scripts on 112 to get results for just the kernels of interests and generate the plots.

Thanks. Will do. Also I still want a full successful CI run to help validate the PR.

> Also, re the paralellism level, wouldn't it make more sense to, in `miirLowerBin` and its friends, have a thing that'll catch exceptions (and _maybe_ signals, though we could just compile with assertions off for release and be under the assumption that we won't segfault) and return failure if an exception occurs? I'm pretty sure that's a typical pattern when wrapping C++ for C?

The crash is not happening in MLIR, but inside MIOpen, which we have less control. In MLIR side within `miirLowerBin` we already properly catch the failures and return the result to MIOpen.

krzysz00 commented:
@whchung Ok, so, to make sure I understand correctly: if compilation fails, either in our code or in LLVM, `miirLowerBin()` will correctly return a failure result to MIOpen and will not crash the process. However, MIOpen assumes that `miirLowerBin()` doesn't fail, and then crashes when it encounters that failure result.

If that's a good summary, then I think we need to some combination of
1. Send a PR to MIOpen to have them check for miirLowerBin() failures and not ship anything that could cause such failures until after they pull that change in
2. Identify ways to stop ourselves failing after we've accepted a set of tuning parameters - this is hard and requires being really careful about what the tuning parameters are and how to fall back to slower algorithms (see, for example, my swizzle code, which has this massive list of preconditions before you can start using swizzles and falls back when they're not available)

whchung commented:
> @whchung Ok, so, to make sure I understand correctly: if compilation fails, either in our code or in LLVM, `miirLowerBin()` will correctly return a failure result to MIOpen and will not crash the process. However, MIOpen assumes that `miirLowerBin()` doesn't fail, and then crashes when it encounters that failure result.
> 
> If that's a good summary, then I think we need to some combination of
> 
> 1. Send a PR to MIOpen to have them check for miirLowerBin() failures and not ship anything that could cause such failures until after they pull that change in
> 2. Identify ways to stop ourselves failing after we've accepted a set of tuning parameters - this is hard and requires being really careful about what the tuning parameters are and how to fall back to slower algorithms (see, for example, my swizzle code, which has this massive list of preconditions before you can start using swizzles and falls back when they're not available)

@jerryyin What's your take on 1? You have more experiences working on MIOpen and should know better about the impact of this proposed change.

jerryyin commented:
> The crash is not happening in MLIR, but inside MIOpen, which we have less control. In MLIR side within miirLowerBin we already properly catch the failures and return the result to MIOpen.

I'm not fully convinced that it is MIOpen that crashed due to MLIR's failure. To accurately determine which component does crash I need to rerun the attach a debugger to make sure. However, I do think that having 20 threads all or majority of which all failing the MLIR compilation would push MIOpen to its extreme in handling a condition it is not prepared for. MIOpen assumes that once something is marked as applicable, it must not fail.

> Send a PR to MIOpen to have them check for miirLowerBin() failures and not ship anything that could cause such failures until after they pull that change in

That's on us because MIOpen get the applicability check result from our API `miirLowerTuningParams`. So this basically reduce to that we do the full compilation in the applicability check. This is fine only as a temporary measure but shall not persist as our permanent solution. We still have to figure out exactly what caused these issues.

krzysz00 commented:
@jerryyin To clarify, the general contract of MIOpen is that, if a solver declares that it is applicable for a particular configuration (problem size + set of tuning parameters) it must always successfully generate a kernel for that configuration, and that, if it somehow fails to generate the kernel, things will explode and it's a bug in the solver, right? Or, equivalently, is `isApplicable(config) == true` implying `bin = generateKernel(config); [bin is a runnable binary]`  a core assumption of MIOpen that they can't/won't change?

If so, then I claim we must not merge KPack until we can refine the applicability checks so that kernel generation below the applicability checks never crashes.

If that isn't a core contract of theirs, then we should
a) If we're the crash source, add a better crash-catching shell
b) If MIOpen's the crash source, improve their error handling

jerryyin commented:
> if it somehow fails to generate the kernel, things will explode and it's a bug in the solver, right? 

Correct, it is clearly communicated as it is the core assumption MIOpen's architecture relied on. It is written in their spec document. I can have Artem to send over that if necessary.

> Or, equivalently, is isApplicable(config) == true implying bin = generateKernel(config); [bin is a runnable binary] a core assumption of MIOpen that they can't/won't change?

Also correct, changing this requires some architectural modifications in MIOpen, unless we have a good reason to make MIOpen to tolerate otherwise. But such request will need a much broader discussion includes why the applicability check contract cannot work for us.

krzysz00 commented:
@jerryyin Makes sense, that's what I thought, thanks for the info!

whchung commented:
jenkins: retest this please.

whchung commented:
The workaround introduced in commit 0003add9 could get us through the "Tune MLIR kernels" stage in build 68.

Let me run the performance metric report and compare with the mainline.

Also it's needed to understand why the fastest kernel in fp16 backward data/weight result in incorrect kernels, and decide:
- whether to alter the tuning search space.
- identify/fix potential bugs inside dialect implementation.
- identify/avoid/fix potential bugs inside LLVM (core or AMDGPU backend).

whchung committed:
Reject incorrect loop bounds in xdlops_gemm_v2 lowering.

whchung committed:
Properly disable KPACK in backward convolutions.

whchung committed:
Fix clang-format.

whchung committed:
Revert "XXX FIXME Disable tuning for backward data and backward weight for fp16 type."

This reverts commit 0003add966c646443ed609f13dcf9db97b844691.

whchung committed:
Ignore manually supplied tuning parameters for GPU verification kernels.

Ignore manually supplied tuning parameters via --perf_config for GPU
verification kernels. This ensures "--pv_with_gpu" works with
"--perf_config".

whchung committed:
Reject too wide KPACK values for fp32/fp16/bf16 types.

whchung committed:
Reject KPerBlock < KPack cases.

whchung committed:
XXX FIXME: Ignore KReduction XDLOPS path for now.

These M/NPerBlock combinations will result in lowering errors at tuning.

Temporarily reject them to tame tuning process.

whchung committed:
Revert "Throttle parallel level in the tuning process."

This reverts commit 91e318bf259bed1d397efcfc2bd6b4e323d9e2f7.

whchung committed:
XXX FIXME: Ignore KReduction path only for forward convolution.

In backward convolutions KPACK are temporarily fixed as 1 so they won't
be affected.

whchung committed:
Merge remote-tracking branch 'origin/miopen-dialect' into miopen-dialect-kpack-logic

whchung committed:
Add operator= to CoordTransformBuilder.

whchung committed:
Adopt TransformMapAttr for filter tensor.

whchung committed:
Adopt TransformMapAttr for input tensor.

whchung committed:
Remove unused codes.

whchung committed:
Fix clang-format.

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -1647,13 +1661,61 @@ template <typename T> struct Conv2DRewritePattern : public OpRewritePattern<T> {
+            loc, gemmFilterPad, kpackGemmFilterTransformAttr);
+      }
+    }
+
```
I think this logic can be simplified a lot.
You can, at the very least, go `kpackGemmFilterTransform.passThrough(padGemmFilterTransform.endName(0))` and so on, and otherwise use `endName()` to not have to branch on whether we padded K or not.

whchung replied:
`endName()` API is not used anywhere in the implementation so I didn't use it. Now it's used in the latest logic.

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -1827,6 +1894,55 @@ template <typename T> struct Conv2DRewritePattern : public OpRewritePattern<T> {
+            loc, gemmInputPad, kpackGemmInputTransformAttr);
+      }
+    }
+
```
Same notes re there being ways to avoid this if nest

whchung replied:
logic here has been simplified.

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -1647,13 +1661,61 @@ template <typename T> struct Conv2DRewritePattern : public OpRewritePattern<T> {
+            loc, gemmFilterPad, kpackGemmFilterTransformAttr);
+      }
+    }
+
```
I think there's a much simpler way to do this, using `endName` on `padGemmFilterTransform` (which might just be `gemmFilterTransform`). That way, you can grab the relevant dimension names, whether they're "gemmXPad" or "gemmX" and don't have to have this pile of branches that doesn't add much.

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -1827,6 +1894,55 @@ template <typename T> struct Conv2DRewritePattern : public OpRewritePattern<T> {
+            loc, gemmInputPad, kpackGemmInputTransformAttr);
+      }
+    }
+
```
Same comment as wrt filter.

whchung replied:
the logic has been simplified.

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -2390,85 +2525,152 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<miopen::GridwiseGemm
+        GemmBBlockCopyThreadSliceLengths_GemmN =
+            GemmBBlockCopyNumberDataPerThread /
+            GemmBBlockCopyThreadSliceLengths_GemmK;
+      }
```
As a general question ... what would happen if, instead of having all this `if (kPack > 1)` stuff, you just always added a kPack dimension that was frequently length 1? Would we expect that to lead to worse ISA?

whchung replied:
We don't know that yet. It could be addressed in a future PR to simplify the logic.

krzysz00 replied:
Fair enough, future PR it is - this wasn't a critical complaint, just an observation. Do we want a ticket for the investigation?

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -5787,17 +6615,23 @@ struct XdlopsGemmV2RewritePattern
+      Value valueA;
+      if (KPack > 1) {
+        valueA = emitLoadLogic(
+            ilmkb, loc, op.matrixA().getType().template cast<MemRefType>(),
```
Couldn't this always be emitLoadLogic and it'll drop down to the memref.load in the scalar case?

whchung replied:
I guess it's possible. But I don't want to change this until a full passing CI. I'd do it in a later PR to further clean this up.

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h:
```diff
@@ -911,6 +928,35 @@ class PopulateParamsXDL : public PopulateParamsBase {
+           param.gemmMPerBlock == 64) &&
+          (param.gemmNPerBlock == 16 || param.gemmNPerBlock == 32 ||
+           param.gemmNPerBlock == 64)) {
+        return failure();
```
Are these perhaps victims of https://ontrack-internal.amd.com/browse/SWDEV-318900

whchung replied:
this deserves further investigation. Currently it doesn't appear to be the case.

krzysz00 commented on mlir/lib/Dialect/MIOpen/Transforms/AffixTuningParameters.cpp:
```diff
@@ -57,11 +57,122 @@ void AffixTuningParameters::affixTuningParametersImpl(T &op) {
+  auto inputShape = inputType.getShape();
+
+  // Get shape of output tensor.
+  auto outputType = op.output().getType().template cast<MemRefType>();
```
With how many times we've duplicated this bit in our codebase ... let's have a future PR where we pull it out to some common utility?

whchung replied:
defintely.

krzysz00 reviewed (commented):
(I'm hoping refreshing mid-review didn't leave duplicate comments)

So, overall, if this works, it looks good.

One thing I'd like to see cleaned up a bit is how the KPack transform is created, I left comments above on that.

The other main observation, which I'd say doesn't block the PR but would make a plausible followup ticket - unless we want to do it now ... is that we've got all this branching on whether we have KPack. Maybe the right thing to do would be to always have the KPack dimension but set it to length 1 most of the time.

whchung committed:
Change perf_config to use a valid tuning parameter.

Also start to check KPACK attribute.

whchung committed:
Address review feedback.

whchung committed:
Simplify logic addressing review feedbacks.

krzysz00 commented:
Overall, this all looks good to me - or at least it looks like we have some cleanup items for later but nothing egregious.

I'd poke someone else to check things over once more, but I'm willing to approve if you're sick of waiting.

krzysz00 commented:
(I mean, the fact that we're going to town with disabling certain xdlops cases is concerning, but that could be a future-us issue.)

whchung commented:
> Overall, this all looks good to me - or at least it looks like we have some cleanup items for later but nothing egregious.
> 
> I'd poke someone else to check things over once more, but I'm willing to approve if you're sick of waiting.

Let's wait until we have a complete passing CI before actually doing so.
I have less confidence on how I use the APIs in `BottomUpCTBuilder` and `TransformMapAttr`.
Other than addressing code refactoring changes in PR #469, we'll also have to adopt changes introduced in PR #461 here.

krzysz00 commented:
I can also take the time tomorrow to run this on MI-200 with my "hack around SWDEV-316900" patch applied to get that angle as well.

whchung committed:
Merge remote-tracking branch 'origin/miopen-dialect' into miopen-dialect-kpack-logic

whchung committed:
Fix lowering_blockwise_gemm_v2.mlir after merge.

Upstream version was marked XFAIL. Fix it here.

whchung committed:
Fix lowering_xdlops_gemm_v2.mlir after merge.

Upstream version was marked XFAIL. Fix it here.

whchung committed:
Fix incorrect logic in mcpuMemset5DHalfRandInt introduced in PR#461.

The signature of the function demands random integer values, not floating
point values.

whchung committed:
Remove compile-time warnings.

whchung committed:
Fix issues in e2e cases with padding.

The semantic of BottomUpCTBuilder::above() needs some examination.

whchung committed:
Fix 1 incorrect FileCheck macro.

krzysz00 committed:
Fix perf tests to use miopen-gen (#476)

whchung committed:
Introduce a default copy ctor to CoordTransformsBuilder.

whchung committed:
Reject invalid tuning parameters only when KPACK is in use.

whchung committed:
Make the build order of miopen-gen be deterministic.

Since PR #461, miopen-gen has become a required tool. Make it be built
explicitly in "Build MLIR" stage to prevent non-determinisim.

whchung commented:
@asroy / @jerryyin / @krzysz00 now finally we have an all-passing CI run.

whchung committed:
Align .clang-format with upstream.

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h:
```diff
@@ -911,6 +928,35 @@ class PopulateParamsXDL : public PopulateParamsBase {
+
+    // Reject too wide KPACK values for fp32/fp16/bf16 types.
+    auto dataType = ctx.getDataType();
+    if (dataType.isF32() && param.gemmKPack >= 8) {
```
Shouldn't this be `> 4`?

whchung replied:
Will fix that in a separate cleanup PR.

krzysz00 commented on mlir/lib/Dialect/MIOpen/Transforms/LowerMIOpenOps.cpp:
```diff
@@ -221,7 +221,6 @@ LogicalResult backwardWeightAtomicAdd(miopen::Conv2DBwdWeightOp op,
     outputNames.push_back(outputAttr.getValue());
 
     if (filterAttr.getValue() == "g") {
-      g = filterShape[i];
```
Why leave the empty if statement?

whchung replied:
Will fix that in a separate cleanup PR.

krzysz00 commented on mlir/test/Dialect/MIOpen/lowering_blockwise_gemm_v2.mlir:
```diff
@@ -1,18 +1,24 @@
-func @miopen_blockwise_gemm_v2_two_results(%matrixA : memref<12288xf32, 3>, %matrixB : memref<12288xf32, 3>) -> (vector<32xf32>, vector<32xf32>) {
+#map = affine_map<(d0) -> (d0 + 512)>
+
+func @miopen_blockwise_gemm_v2_two_results(%matrixA : memref<512xf32, 3>, %matrixB : memref<512xf32, #map, 3>, %bufferA : memref<2xvector<2xf32>, 5>, %bufferB : memref<2xvector<2xf32>, 5>) -> (vector<32xf32>, vector<32xf32>) {
```
Re this test: I don't think this'll properly exercise the function we care about. To match how `blockwise_gemm_v2` gets invoked, you'd want to pass in a `Slice` transform_map with `#map` as its backing affine map

whchung replied:
The test was marked XFAIL in `miopen-dialect` for some time. It was originally created to test if a `blockwise_gemm_v2` op could be lowered to `xdlops_gemm_v2` successfully. We can improve the test, and also `lowering_xdlops_gemm_v2.mlir` to match more closely to how we really use these ops. Will do that in a separate cleanup PR.

whchung replied:
This test is actually already reflecting how we use `blockwise_gemm_v2`. The second memref does carry an affine map which is the result of `Slice`.

krzysz00 commented on mlir/test/Dialect/MIOpen/lowering_xdlops_gemm_v2.mlir:
```diff
@@ -1,18 +1,25 @@
+    n_per_wave = 64 : i32,
+    n_waves = 2 : i32,
+    transforms = [[], []]
+  } : memref<512xf32, 3>, memref<512xf32, #map, 3>, index, index, memref<2xvector<2xf32>, 5>, memref<2xvector<2xf32>, 5>, vector<32xf32>, vector<32xf32> -> vector<32xf32>, vector<32xf32>
```
Same comment re raw affine maps on the memref

whchung replied:
Will do that in a separate cleanup PR.

whchung replied:
Same as comment above.

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -1650,13 +1665,40 @@ template <typename T> struct Conv2DRewritePattern : public OpRewritePattern<T> {
+      kpackGemmFilterTransform.passThrough(sourceTransform.endName(2));
+      // Use Unmerge to split gemmK (dim 1) into gemmK and gemmKPack, place
+      // gemmKPack at dim 3.
+      int64_t gemmKLength = sourceTransform.endSize(sourceTransform.endName(1));
```
Nit: We could probably do `kName = sourceTransform.endName(1);` before this bit

whchung replied:
Will fix that in a separate cleanup PR.

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -1824,12 +1870,40 @@ template <typename T> struct Conv2DRewritePattern : public OpRewritePattern<T> {
+      kpackGemmInputTransform.passThrough(sourceTransform.endName(2));
+      // Use Unmerge to split gemmK (dim 1) into gemmK and gemmKPack, place
+      // gemmKPack at dim 3.
+      int64_t gemmKLength = sourceTransform.endSize(sourceTransform.endName(1));
```
Did I not add an `endSize()` that takes an index` making this `sourceTransform.endSize(1);`? If not ... whoops.

whchung replied:
There is indeed an `endSize(uint32_t dim)` interface. I was just use all names here. Will fix that in a separate cleanup PR.

krzysz00 reviewed (approved):
Overall, this looks like a quite solid first pass at kPack support.

I've left some minor nitpicks, but none of those are serious enough that I'd make them block merging this.

@whchung, since this PR has a lot of "TODO"/"add support for" sections, please create tickets for them. For example,
- non-xdlops kpack
- Adding kPack to backwards ops where possible (wrw seems like a solid candidate, backward data is rather agressively scalar in its memory reads)
- Potentially tightening up some of the logic so there's less weird copy-paste

Thanks for suffering through all my comments and for putting in all the work needed to get these massive speedups!
